File: 300-LongestIncreasingSubsequence/300-LongestIncreasingSubsequence.java
// Last updated: 2/27/2026, 5:19:25 PM
class Solution {
    public int lengthOfLIS(int[] nums) {
        
        int[] curr = new int[nums.length + 1];
        int[] next = new int[nums.length + 1];

        for(int index = nums.length - 1; index >= 0 ; index--){
            for(int prev = index - 1; prev >= -1; prev--){

                int len = 0 + next[prev+1];

                if(prev == -1 || nums[index] > nums[prev]){
                    len = Math.max(len, 1 + next[index + 1]);
                }

                curr[prev + 1] = len;
            }

            next = curr;
        }
        
        return next[-1+1];
    }
}

  

File: 300-LongestIncreasingSubsequence/300-LongestIncreasingSubsequence.py
# Last updated: 7/12/2026, 6:20:18 PM
class Solution:
    def lengthOfLIS(self, nums):
        tails = []

        for num in nums:
            left, right = 0, len(tails)

            while left < right:
                mid = (left + right) // 2

                if tails[mid] < num:
                    left = mid + 1
                else:
                    right = mid
          
            if left == len(tails):
                tails.append(num)
            else:
                tails[left] = num

        return len(tails)